src/cljc/rems/common/form.cljc

```diff
@@ -62,3 +79,297 @@
+
+         :else
+         (let [parsed (parse-int max-length)]
+           (when (or (nil? parsed)
+                     (not (<= (first max-length-range) parsed (second max-length-range))))
+             :t.form.validation/invalid-value)))})
+
+(defn- validate-option [option id languages]
+  {id (merge (validate-text-field option :key)
+             (validate-localized-text-field option :label languages))})
+
+(defn- validate-options [options languages]
+  {:field/options (apply merge (mapv #(validate-option %1 %2 languages) options (range)))})
+
+(defn- field-option-keys [field]
+  (set (map :key (:field/options field))))
+
+(defn- validate-only-if-field [field visibility fields]
+  (let [referred-id (get-in visibility [:visibility/field :field/id])
+        referred-field (find-first (comp #{referred-id} :field/id) fields)]
+    (cond
+      (not (:visibility/field visibility))
+      {:field/visibility {:visibility/field :t.form.validation/required}}
+
+      (empty? referred-id)
+      {:field/visibility {:visibility/field :t.form.validation/required}}
+
+      (not referred-field)
+      {:field/visibility {:visibility/field :t.form.validation/invalid-value}}
+
+      (not (supports-options? referred-field))
+      {:field/visibility {:visibility/field :t.form.validation/invalid-value}}
+
+      (empty? (:visibility/values visibility))
+      {:field/visibility {:visibility/values :t.form.validation/required}}
+
+      (some #(not (contains? (field-option-keys referred-field) %)) (:visibility/values visibility))
+      {:field/visibility {:visibility/values :t.form.validation/invalid-value}})))
+
+(defn- validate-visibility [field fields]
+  (when-let [visibility (:field/visibility field)]
+    (case (:visibility/type visibility)
+      :always nil
+      :only-if (validate-only-if-field field visibility fields)
+      nil {:field/visibility {:visibility/type :t.form.validation/required}}
+      {:field/visibility {:visibility/type :t.form.validation/invalid-value}})))
+
+(defn- validate-fields [fields languages]
+  (letfn [(validate-field [index field]
+            {index (merge (validate-field-type field)
+                          (validate-localized-text-field field :field/title languages)
+                          (when (supports-placeholder? field)
+                            (validate-optional-localized-field field :field/placeholder languages))
+                          (when (supports-max-length? field)
+                            (validate-max-length (:field/max-length field)))
+                          (when (supports-options? field)
+                            (validate-options (:field/options field) languages))
+                          (when (supports-visibility? field)
+                            (validate-visibility field fields)))})]
+    (apply merge (map-indexed validate-field fields))))
+
+(defn- nil-if-empty [m]
+  (when-not (empty? m)
+    m))
+
+(defn validate-form [form languages]
```

Comment:
should it be called validate-form-template to avoid confusion with rems.form-validation?